### DISCUSSION TITLE
How to deal with features in diesel

### DISCUSSION THREAD

**Discussion by NoxTav** [diesel-rs/diesel#4377] [Q&A]
### Versions

- **Rust:** 1.81.0
- **Diesel:** 2.2.4
- **Database:** sqlite
- **Operating System** Ubuntu 24

### Feature Flags

- **diesel:**  sqlite, uuid, r2d2, returning_clauses_for_sqlite_3_35, chrono

### What are you trying to accomplish?

It is kind of related to https://github.com/diesel-rs/diesel/issues/860, but not at the same level.

We have an API that uses the same structure to insert, update, delete or select. We usually need to return all fields of the table. That work well. But we have a model (and maybe more later) that uses features to not use certain fields. 

We return the model in all cases. 

We use a trait to implement functions like insert, select, update and delete on the model.

But the trait seems unable to compile. But when the function is in an implementation, it compiles.

*Is there some best advice to deal with features in the structures?*

Or if there isn't, is there something wrong with the way we do our traits? Is there an obvious missing trait?

### What is the expected output?

Compiling using a trait.

### What is the actual output?

Our trait don't compile because of impossible requirements, for example, SingleValue to return several fields.

### Steps to reproduce

See https://gist.github.com/NoemieTav/bce7a200fa07b8ee39810fa13b09290f for example.

The trait is commented, 

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) and the [discussion forum](https://github.com/diesel-rs/diesel/discussions) for similar possible closed issues.
   - https://github.com/diesel-rs/diesel/issues/860
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment]
I converted this to a discussion as this is not an issue in diesel but a support question. Please do not fill such questions as bug reports. I will just close the next one as invalid, instead of converting them.

To the actual question: I will try to write an answer later

**Comment by weiznich** [comment]
For reference the provided example fails to compile with the following error message:

```
error[E0412]: cannot find type `custom_attributes` in module `crate::schema::node_extensions`
  --> src/main.rs:87:9
   |
87 |     pub custom_attributes: Option<String>,
   |         ^^^^^^^^^^^^^^^^^ not found in `crate::schema::node_extensions`

error[E0425]: cannot find value `custom_attributes` in module `crate::schema::node_extensions`
  --> src/main.rs:87:9
   |
87 |     pub custom_attributes: Option<String>,
   |         ^^^^^^^^^^^^^^^^^ not found in `crate::schema::node_extensions`

```

That's expected as there is no `node_extension` column in your database schema. You either need to add it there or you need to specify how else this field should be handled for insertion and selection.

For insertion it's possible to ignore it via
```rust
    #[diesel(skip_insertion)]
```

For selection it's possible to select a dummy value instead via
```rust
    #[diesel(select_expression = None::<String>.into_sql::<Nullable<Text>>())]
    #[diesel(select_expression_type = diesel::dsl::IntoSql<Option<String>, Nullable<Text>>)]
```

More generally speaking: It's usually preferred to just split the structs if you don't need all fields in all cases, that's much better than working with configs or fake values.

**Reply by JMLX42** [reply]
> Node::insert(&database, vec![child_node.clone(), root_node.clone()]).unwrap();

@weiznich you mean split the model struct without splitting the actual SQL table?

Can you please give an example?

**Reply by weiznich** [reply]
@JMLX42 I generally do not answer questions if I'm pinged without previous interactions.
The only exceptions are significant sponsors of diesel and someone that reports a security vulnerability. Neither is the case here so you are on your own.

**Reply by JMLX42** [reply]
@weiznich not even to contributors? :cry:

**Reply by weiznich** [reply]
Yes, that also applies to contributors at least as long as they have not a history of long and significant contributions. I mean there are a few 100 diesel contributors out there (and a lot more users), I just don't have the capacity to be pinged by all of them so I need to make hard rules. If you have questions you can always ask questions. I usually will answer them when I have capacity (and if I'm not pinged).
